Remove extra space from command name

bbockelm opened the pull request:
When `$CMD` was switched to a bash list instead of a string, bash no longer stripped off the extra '` `' character.  Thus, Singularity was attempting to execute a binary named `' foo'` instead of `'foo'`.

I was able to recreate this by logging in to a worker node with a running glidein, making a copy of the job wrapper script, and applying the patch by hand.  I was able to:
* Reproduce the errors reported by Diego in https://ggus.eu/index.php?mode=ticket_info&ticket_id=128469,
* Verify that this fix took care of the reported error.
* Verify that arguments with spaces in them appear to work.

Mystery solved, hopefully!

@rynge - it's not clear to me why the OSG pilots don't hit the same issue.  Thoughts?

bbockelm committed:
Remove extra space from command name

When `$CMD` was switched to a bash list instead of a string, bash no longer stripped off the extra '` `' character.  Thus, Singularity was attempting to execute a binary named ` foo` instead of `foo`.

I was able to recreate this by logging in to a worker node with a running glidein, making a copy of the job wrapper script, and applying the patch by hand.  I was able to:
* Reproduce the errors reported by Diego in https://ggus.eu/index.php?mode=ticket_info&ticket_id=128469,
* Verify that this fix took care of the reported error.
* Verify that arguments with spaces in them appear to work.

Mystery solved, hopefully!

rynge commented:
I don't think that will work if users have arguments with dashes, as those will be interpreted by echo and it will error out. I don't know why I missed that part in my PR, but OSG VO has the space, and then it is removed at the end of the pipe:

        # build a new command line, with updated paths
        CMD=()
        for VAR in "$@"; do
            # Two seds to make sure we catch variations of the iwd,
            # including symlinked ones. The leading space is to prevent
            # echo to interpret dashes.
            VAR=`echo " $VAR" | sed -E "s;$PWD(.*);/srv\1;" | sed -E "s;.*/execute/dir_[0-9a-zA-Z]*(.*);/srv\1;" | sed -E "s;^ ;;"`
            CMD+=("$VAR")
        done

bbockelm committed:
Copy extra `sed` invocation from OSG

Utilize the same track that OSG does to avoid problems with command line arguments: add an extra space at the beginning of the argument to foil `echo`, then remove it later on using `sed`.

bbockelm commented:
@rynge - how does it look now?

rynge commented:
Looks good.
